pandemic51/core/streaming.py

```diff
@@ -245,35 +111,420 @@ def sample_first_frame(inpath, outpath, width=None, height=None):
+            dt: datetime object of when the image was downloaded
+        '''
+        raise NotImplementedError("Subclass must implement")
+
+    def download_image_and_store(self, outdir):
+        '''Downloads an image from the latest stream, and add it to the
+        database.
+
+        If an image with the same filename was already on disk, no new entry is
+        added to the database.
+
+        Args:
+            stream_name: the stream name
+            outdir: the output directory
+
+        Returns:
+            image_path: path the the downloaded image on disk
+            dt: datetime object of when the image was downloaded
+        '''
+        is_new_img, image_path, dt = self.download_image(outdir)
+
+        if is_new_img:
+            add_stream_history(self.stream_name, dt, image_path)
+
+        return image_path, dt
+
+    @classmethod
+    def from_dict(cls, d, *args, **kwargs):
+        '''Constructs a Stream object from a JSON dictionary.
+
+        Args:
+            d: a JSON dictionary representation of a Serializable object
+
+        Returns:
+            a Stream instance
+        '''
+        downloader_cls = etau.get_class(d["type"])
+        return downloader_cls._from_dict(d)
+
+    @classmethod
+    def from_json(cls, path, *args, **kwargs):
+        '''Constructs a Stream object from a JSON file.
+
+        If `stream_name` key is missing, it is automatically populated.
+
+        Args:
+            path: the path to the JSON file on disk
+
+        Returns:
+            a Stream instance
+        '''
+        d = etas.read_json(path)
 
-    Returns:
-        image_path: path the the downloaded image on disk
-        dt: datetime object of when the image was downloaded
+        if "stream_name" not in d:
+            d["stream_name"] = os.path.splitext(os.path.basename(path))[0]
+
+        return cls.from_dict(d, *args, **kwargs)
+
+    @classmethod
+    def from_stream_name(cls, stream_name):
+        '''Constructs a Stream object.
+
+        Args:
+            stream_name: the name of the stream
+
+        Returns:
+            a Stream instance
+        '''
+        return cls.from_json(cls.stream_path(stream_name))
+
+    @classmethod
+    def _from_dict(cls, d):
+        raise NotImplementedError("Subclass must implement")
+
+
+class M3U8Stream(Stream):
+    '''A Stream class that reads URIs from an M3U8 chunk path'''
+    def __init__(self, stream_name, GMT, webpage, chunk_path):
+        super(M3U8Stream, self).__init__(stream_name, GMT)
+        self.webpage = webpage
+        self.chunk_path = chunk_path
+
+    def get_live_stream_url(self):
+        url = self.chunk_path
+
+        try:
+            urllib.request.urlopen(url)
+        except urllib.error.HTTPError:
+            self.update_stream_chunk_path()
+            url = self.chunk_path
+
+        if "videos2archives" in url:
+            url = (
+                    "https://pdi-service.voxel51.com/stream-archive/" +
```

Comment:
Consider refactoring these magic URLs to config constants? Others have been done like that